Sources/InteroperabilityTests/InteroperabilityTestCases.swift

```diff
@@ -0,0 +1,721 @@
+    try assertEqual(metadata.count, 1)
+    var metadataIterator = metadata.makeIterator()
+    guard let trailingMetadataPair = metadataIterator.next() else {
+      try assertFailure("The trailing metadata should contain a key value pair.")
+      return
+    }
+    try assertEqual(trailingMetadataPair.key, self.trailingMetadataName)
+    try assertEqual(trailingMetadataPair.value, Metadata.Value.binary(self.trailingMetadataValue))
+  }
+
+  func run(client: GRPCClient) async throws {
+    let testServiceClient = Grpc_Testing_TestService.Client(client: client)
+
+    let unaryRequest = Grpc_Testing_SimpleRequest.with { request in
+      request.responseSize = 314_159
+      request.payload = Grpc_Testing_Payload.with {
+        $0.body = Data(count: 271_828)
+      }
+    }
+    var metadata = Metadata()
+    metadata.addString(self.initialMetadataValue, forKey: self.initialMetadataName)
+    metadata.addBinary(self.trailingMetadataValue, forKey: self.trailingMetadataName)
+    let streamingMetadata = metadata
+
+    try await testServiceClient.unaryCall(
+      request: ClientRequest.Single(message: unaryRequest, metadata: metadata)
+    ) { response in
+      // Check the initial metadata.
+      let receivedInitialMetadata = response.metadata
+      try checkInitialMetadata(receivedInitialMetadata)
+
+      // Check the message.
+      try assertEqual(response.message.payload.body, Data(count: 314_159))
+
+      // Check the trailing metadata.
+      try checkTrailingMetadata(response.trailingMetadata)
+    }
+
+    let streamingRequest = ClientRequest.Stream(metadata: streamingMetadata) { writer in
+      let message = Grpc_Testing_StreamingOutputCallRequest.with {
+        $0.responseParameters = [
+          Grpc_Testing_ResponseParameters.with {
+            $0.size = 314_159
+          }
+        ]
+        $0.payload = Grpc_Testing_Payload.with {
+          $0.body = Data(count: 271_828)
+        }
+      }
+      try await writer.write(message)
+    }
+
+    try await testServiceClient.fullDuplexCall(request: streamingRequest) { response in
+      switch response.accepted {
+      case .success(let contents):
+        // Check the initial metadata.
+        let receivedInitialMetadata = response.metadata
+        try self.checkInitialMetadata(receivedInitialMetadata)
+
+        var responseParts = contents.bodyParts.makeAsyncIterator()
+        for _ in 1 ... 2 {
+          switch try await responseParts.next() {
+          // Check the message.
+          case .message(let message):
+            try assertEqual(message.payload.body, Data(count: 314_159))
+          // Check the trailing metadata.
+          case .trailingMetadata(let receivedTrailingMetadata):
+            try self.checkTrailingMetadata(receivedTrailingMetadata)
+          default:
+            try assertFailure(
+              "The client should have received only a message and trailing metadata."
+            )
+          }
+        }
```

Comment:
This only validates that we receive at least two parts, not exactly two.

One trick we can do is collect all of the parts: 

```swift
let parts = try await responseParts.reduce(into: []) { $0.append($1) }
try assertEqual(parts.count, 2)
// ...

```